lidofinance/ethereum-staking-widget
# feat: IPFS mode, RPC settings page and refactoring

### Description

Add IPFS mode:
- SPA routing through hash.
- Fix using a NextLink for IPFS mode (also needs for a migration process to Next v13 when we will gonna).
- HMR for runtime of `dev-ipfs`.
- Page settings.
- Hide the statistics block if all stats items not available.
- New ENV variables.
- CSP for IPFS mode.


### Code review notes

Pay attention:
- Change behavior of `LocalLink `.
- Check `LinkIpfs` component.
- Splitted `index page` to`home/home-page-ipfs.tsx` and to `pages/home/home-page-regular.tsx`.
- Check using `dynamics.ipfsMode` from `public/runtime/window-env.js` instead of `IPFS_MODE` from `publicRuntimeConfig`.
- Using swr STRATEGY_LAZY.
- Body of useGetRpcUrlByChainId func.
- CSP for IPFS.

### Testing notes

Needs to check:
- All links surfing (in IPFS mode and regular mode).
- **NOT** sending matomo events (but technically it's prepared).
- Change RPC in settings page.
- Main functions like: connect wallet, stake, wrap and etc.
- Hiding the statistics block.
- Settings page (to change RPC and check RPC related functional)
- Check that all native link operations haven't broken (like "open in a new window").

Also needs to check a working on different IPFS gateways (see: https://ipfs.github.io/public-gateway-checker/)

[Free public Mainnet RPC you can find here](https://chainlist.org/chain/1)
[Free public Goerli RPC you can find here](https://chainlist.org/chain/5)

### Checklist:
